### ISSUE TITLE
VS Code still have to access internet when installing extensions with local VSIX file

### ISSUE THREAD

**Issue by hsinyinfu** [microsoft/vscode#142812]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.63.2
- OS Version: Windows 10 20H2 and Ubuntu 18.04.5 (with remote-ssh scenario)

Steps to Reproduce:

- Scenario A: Install extension with a VSIX file using command line tool on Windows 10 or Ubuntu 18.04.5
1. Prepare an extension vsix file (e.g., Prettier or Python)
2. Use the following command to install extension with the vsix file
```
code --install-extension <path to the vsix file>
```

- Scenario B: Install extension with a VSIX file using VS Code UI on Windows 10
1. Prepare an extension vsix file (e.g., Prettier or Python)
2. In the "Extensions" view, click "Install from VSIX" and select the vsix file to be installed.

- Scenario C: Install extension with a VSIX file using VS Code UI on remote Ubuntu Server (remote-ssh activated)
1. Setup remote-ssh extension and environment settings
2. Connect to the remote Ubuntu server
3. In the remote-ssh VS Code window, click "Install from VSIX" and select the vsix file on the remote Ubuntu server to be installed.

Problem States:

In my company network, both local Windows PC and remote Ubuntu Linux terminal server are not allowed to access internet with firewall settings blocked.
Therefore, I downloaded the extension vsix files in my home and sent them to my company PC and installed them with the steps mentioned above.
However, I found that when installing extensions from vsix files, before VS Code start to extract contents from the vsix files, VS Code still tried to access the Microsoft marketplace, which makes me waste a lot of time (about 2minutes per installation) waiting VS Code request timeout and then start to do the extraction and installation.
In my understanding, all the content needed for installing an extension are all included in the vsix file. 
**Why do VS Code still need to send POST request to marketplace every time at the beginning when installing from vsix file?**
Or, if is it possible to **provide a configuration option to disable this behavior, or at least make users have a chance to set the timeout value for requests sent by VS Code to Microsoft marketplace?**

Here is the log when installing extension with vsix file on my PC which can access internet (I set `--log trace`),
as you can see, it accessed the marketplace in the log 
`TRACE RequestService#request https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery`
```
TRACE Started scanning user extensions
 INFO CLI main {
  _: [],
  diff: false,
  add: false,
  goto: false,
  'new-window': false,
  'reuse-window': false,
  wait: false,
  help: false,
  'list-extensions': false,
  'show-versions': false,
  'install-extension': [
    'C:\\Users\\mtk23649.DOMAIN_MTK\\Desktop\\VSCode Portable\\extensions\\esbenp.prettier-vscode\\esbenp.prettier-vscode-9.1.0.vsix'
  ],
  'pre-release': false,
  version: false,
  verbose: false,
  log: 'trace',
  status: false,
  'prof-startup': false,
  'no-cached-data': false,
  'prof-v8-extensions': false,
  'disable-extensions': false,
  'disable-gpu': false,
  'ms-enable-electron-run-as-node': false,
  telemetry: false,
  debugRenderer: false,
  logExtensionHostCommunication: false,
  'skip-release-notes': false,
  'skip-welcome': false,
  'disable-telemetry': false,
  'disable-updates': false,
  'disable-keytar': false,
  'disable-workspace-trust': false,
  'disable-crash-reporter': false,
  'skip-add-to-recently-opened': false,
  'unity-launch': false,
  'open-url': false,
  'file-write': false,
  'file-chmod': false,
  'driver-verbose': false,
  force: false,
  'do-not-sync': false,
  trace: false,
  'force-user-env': false,
  'force-disable-user-env': false,
  'open-devtools': false,
  __sandbox: false,
  'no-proxy-server': false,
  'no-sandbox': false,
  nolazy: false,
  'force-renderer-accessibility': false,
  'ignore-certificate-errors': false,
  'allow-insecure-localhost': false,
  logsPath: 'C:\\Users\\mtk23649.DOMAIN_MTK\\Desktop\\VSCode Portable\\stables\\VSCode-win32-x64-1.63.2\\data\\user-data\\logs\\20220211T152929'
}
Installing extensions...
TRACE Started scanning user extensions
TRACE Scanned user extensions: 4
TRACE Scanned user extensions: 4
(node:28292) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `Code --trace-deprecation ...` to show where the warning was created)
TRACE Started scanning user extensions
TRACE Scanned user extensions: 4
TRACE ExtensionManagementService#install file:///c%3A/Users/mtk23649.DOMAIN_MTK/Desktop/VSCode%20Portable/extensions/esbenp.prettier-vscode/esbenp.prettier-vscode-9.1.0.vsix
 INFO Installing extension: esbenp.prettier-vscode
TRACE Started scanning system extensions
TRACE Started scanning user extensions
TRACE Scanned user extensions: 4
TRACE Scanned system extensions: 86
TRACE Started scanning system extensions
TRACE Started scanning user extensions
TRACE Scanned user extensions: 4
TRACE Scanned system extensions: 86
TRACE Started scanning user extensions
TRACE Scanned user extensions: 4
TRACE RequestService#request https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery
TRACE resolveShellEnv(): skipped (Windows)
TRACE Started extracting the extension from c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\extensions\esbenp.prettier-vscode\esbenp.prettier-vscode-9.1.0.vsix to C:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [ADDED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
TRACE [File Watcher (node.js)] [CHANGED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224
 INFO Extracted extension to C:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\.3be0a99b-f662-4b55-8289-5651a1ebb224: esbenp.prettier-vscode
TRACE [File Watcher (node.js)] [ADDED] c:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\esbenp.prettier-vscode-9.1.0
 INFO Renamed to C:\Users\mtk23649.DOMAIN_MTK\Desktop\VSCode Portable\stables\VSCode-win32-x64-1.63.2\data\extensions\esbenp.prettier-vscode-9.1.0
 INFO Extracting completed. esbenp.prettier-vscode
 INFO Extension installed successfully: esbenp.prettier-vscode
Extension 'esbenp.prettier-vscode-9.1.0.vsix' was successfully installed.
```

**Comment by sandy081** [comment]
Yeah, we reach to Marketplace to update the metadata of the extension. We can think about having an offline option.

**Comment by jsjoeio** [comment]
I'm curious how this works under the hood and how difficult it would be to change. Gonna leave my notes below:

## Notes

The `installVSIX` method: https://github.com/microsoft/vscode/blob/f74d2c0f62276e9d90f37a00b732a8b7fffd57d4/src/vs/platform/extensionManagement/common/extensionManagementCLIService.ts#L169-L192

I think it is downloading the extension from the gallery to validate the manifest instead of validating the one passed in and that's causing the network request.

Hoping someone can confirm/deny if this is true.

**Comment by sandy081** [comment]
> I think it is downloading the extension from the gallery to validate the manifest instead of validating the one passed in and that's causing the network request.

I do not think so. In the above code, we are getting the manifest from the vsix and validate it.

**Comment by sandy081** [comment]
We reach to Marketplace to get metadata of the extension which is here

https://github.com/microsoft/vscode/blob/1a7d1b4230ea418f2261b1efe716513bca1d7b32/src/vs/platform/extensionManagement/node/extensionManagementService.ts#L708

**Comment by jsjoeio** [comment]
Looks like that function is called here: 

https://github.com/microsoft/vscode/blob/1a7d1b4230ea418f2261b1efe716513bca1d7b32/src/vs/platform/extensionManagement/node/extensionManagementService.ts#L673

Maybe I can this function [`doRun`](https://github.com/microsoft/vscode/blob/1a7d1b4230ea418f2261b1efe716513bca1d7b32/src/vs/platform/extensionManagement/node/extensionManagementService.ts#L669) to return `Promise<{ local: ILocalExtension; metadata?: Metadata }>` instead or something.

**Comment by jsjoeio** [comment]
@hsinyinfu have you tried with the latest version of Code? I couldn't reproduce it.

**Comment by hsinyinfu** [comment]
Hi @jsjoeio,
I just tried to install a hello-world extension I wrote with local vsix file in the latest 1.70.2 version of VS Code,
and it still took 2 minutes and 10 seconds to install it.

The experiments were conducted on my desktop in my office. 
I started VS Code 1.70.2 on that desktop (Windows OS), used Remote-SSH to connect to the remote Ubuntu 18.04 server provided by our IT, and installed that hello-world extension in the Remote-SSH context.
The remote Ubuntu server is forbidden to access internet in the firewall policies.

In addition, I tested the installation on the same machine and the remote server with VS Code 1.69.2 and 1.63.2.
All of them took 2mins and 10s to install this tiny sample extension with local vsix file.
FYI.

Thanks.

**Comment by jsjoeio** [comment]
> installed that hello-world extension in the Remote-SSH context.
The remote Ubuntu server is forbidden to access internet in the firewall policies.

So it worked then, right? 

Or you're saying it takes 2mins and 10s because of some logic trying to connect to the internet still?

**Comment by hsinyinfu** [comment]
Yes, I'm saying that I think VS Code took so much time installing an extension offline from a vsix file because it reaches out the Microsoft marketplace first before starting extracting contents from the vsix file. And even in the latest 1.70.2 version, the problem is still there.

![image](https://user-images.githubusercontent.com/15996457/186732656-2dd305b9-ba97-4892-9888-2cd2cc450ffa.png)

As you can see in the screenshot above, there is a time gap for about 2 minutes bettwen the log "Installing extension" and the log "Extracted extension".
Also, in the Network view of the built-in Developer tool, there are two extensionquery events, one failed, and the other failed due to connection timeout.

In addition, I screenshoted the log of VS Code during installing this extension offline (set log level to `trace`) shown below.
Most time was spent on the request to `https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery` (where I marked with a yellow rectangle).
![image](https://user-images.githubusercontent.com/15996457/186736131-5612f51b-60a1-4feb-b942-39e9a1ad1b0c.png)

To be clear, the offline installation always succeeds. However, the main problem is that it takes too much time trying to reach out the Microsoft marketplace, which is not necessary in my opinion.

**Comment by jsjoeio** [comment]
> To be clear, the offline installation always succeeds. However, the main problem is that it takes too much time trying to reach out the Microsoft marketplace, which is not necessary in my opinion.

Ahhh okay. This makes much more sense. I turned off my internet and installed an extension via a `vsix` file. It completed very quickly so I think the behavior might be different in an offline vs limited-access environment.

Does that mirror what you're seeing?

**Comment by hsinyinfu** [comment]
Yes,  I also tested installing from a `vsix` file when my "personal PC" is completed out of internet by removing the ethernet cable from the PC (no remote-ssh to remote Linux server, and no any firewall policy), and it did complete very quickly.
So you're right, the behavior is different in an complete offline (no any network access) vs limited-access (e.g., the request is blocked by outbound firewall policy) environment.

![image](https://user-images.githubusercontent.com/15996457/187256642-bbf966b0-b667-438b-bdc5-5ba8d4062d62.png)

**Comment by hsinyinfu** [comment]
Hi, is there any update about this issue?

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!

**Comment by sandy081** [comment]
Installing from VSIX behaviour has changed now that reaching out to Marketplace no longer blocks installation.
